src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala

```diff
@@ -0,0 +1,216 @@
+    val lines = Io.readLinesFromResource(resource)
+    val path  = Io.makeTempFile("script.", suffix = Suffix)
+    path.toFile.deleteOnExit()
+    Io.writeLines(path, lines)
+    path
+  }
+}
+
+/** Defines values used to get the version of the executable. */
+trait Versioned {
+  self: CommandLineTool =>
+
+  /** The default version flag. */
+  val VersionFlag: String   = "--version"
+
+  /** Argument used to test version of the executable. */
+  val TestArgs: Seq[String] = Seq(VersionFlag)
+
+  /** Returns version of this executable. */
+  lazy val Version: Try[ListBuffer[String]] = execCommand(Executable +: TestArgs: _*)
+}
+
+/** Defines methods used to check if specific modules are installed with the executable . */
+trait Modular {
+  self: CommandLineTool =>
+
+  /** Command used to test if a module is included with the tested executable. */
+  def TestModuleCommand(module: String): Seq[String]
+
+  /** Command used to test if multiple modules are included with the tested executable. */
+  def TestModuleCommand(modules: String*): Seq[Seq[String]] = modules.map(TestModuleCommand)
+
+  /** Returns true if the tested module exist with the tested executable. */
+  def ModuleAvailable(module: String): Boolean = execCommand(TestModuleCommand(module): _*).isSuccess
+
+  /** Returns true if all tested modules exist with the tested executable.
+    *
+    * For example:
+    * {{
+    * scala> import com.fulcrumgenomics.commons.io._
+    * scala> Rscript.ModuleAvailable(Seq("ggplot2", "dplyr"))
+    * res1: Boolean = true
+    * }}
+    */
+  def ModuleAvailable(modules: Seq[String]): Boolean =  modules.par.map(ModuleAvailable).forall(_ == true)
+}
+
+/** A collection of values and methods specific for the Rscript executable */
+object Rscript extends CommandLineTool with Versioned with Modular with ScriptRunner {
+  val Executable: String     = "Rscript"
+  val Suffix: FilenameSuffix = ".R"
+
+  /** Command to test if a R module exists.
+    *
+    * @param module name of the module to be tested
+    * @return command used to test if a given R module is installed
+    * */
+  def TestModuleCommand(module: String): Seq[String] = Seq(Executable, "-e", s"stopifnot(require('$module'))")
+
+  /** Only returns true if Rscript exists. */
+  override lazy val Available: Boolean = execCommand(Executable, VersionFlag).isSuccess
+
+  /** Only returns true if both Rscript exists and the library ggplot2 is installed. */
+  lazy val ggplot2Available: Boolean = Available && ModuleAvailable("ggplot2")
+}
+
+/** Defines tools to test various version of python executables */
+trait Python extends CommandLineTool with Versioned with Modular with ScriptRunner {
+  val Executable: String
+  val Suffix: FilenameSuffix = ".py"
+  def TestModuleCommand(module: String): Seq[String] = Seq(Executable, "-c", s"import $module")
+}
+
+/** Objects specific for different Python versions.  */
+object Python extends Python  { val Executable: String = "python" }
```

Comment:
how about an apply method instead?
```scala
object Python {
  def apply(version: String = ""): Python = new Python { val Executable: String = f"python$version" }
  val Python2 = Python("2")
  val Python3 = Python("3")
}
```

Reply:
For this I changed the original Python `trait` to a `class` to make it work

Reply:
Been going back and forth with this. As originally I changed this block of code to according to @nh13 's suggestion:
```
/** Defines tools to test various version of python executables */
class Python (val pythonVersion: String, val executable: String = "python") extends
  CommandLineTool with Versioned with Modular with ScriptRunner {
  val suffix: FilenameSuffix = ".py"
  def TestModuleCommand(module: String): Seq[String] = Seq(executable, "-c", s"import $module")
}

/** Objects specific for different Python versions.  */
object Python {
  def apply(pythonVersion : String = "", executable: String = "python" ) =
    new Python(pythonVersion, f"$executable$pythonVersion")
  val Python2 = new Python("2")
  val Python3 = new Python("3")
}
```
However, there is concern that we probably want `Python` to be singleton. Therefore this is changed back to `trait` at the moment. Major `Python` versions are traits extended from `Python`.